### PR TITLE
(Develop) BioEngineer layer for ZeroShot JacobianContact in place.

### PR DESCRIPTION
**Pull request by alijosephine**
Part of integrating categorical jacobian based contact map task into biotrainer-autoeval framework.
Completed the BioEngineer layer. Next up is the AutoEval layer.

### CODE REVIEW

File: biotrainer/bioengineer/bioengineer.py [hunk 1]
```diff
@@ -237,3 +240,86 @@ def rank_variant_scores(variant_scores: List[VariantScore], actual_scores: Dict[
...
+        """
+        Derive contact map from categorical Jacobian.
+
+        Returns:
+            np.ndarray: [L, L]
+        """
+        return self.model_wrapper.zero_shot_contact_map(sequence, batch_size)
+
+    def run_contact_dataset(self,
+                          dataset_dir_path: Union[str, Path],
```
Comment (SebieF) [inline biotrainer/bioengineer/bioengineer.py]: I would suggest to replace that by a direct input parameters for the fasta file and the .npy files. This way, we can be agnostic of the exact required directory structure.

File: biotrainer/bioengineer/bioengineer_custom_model.py [hunk 2]
```diff
@@ -66,6 +66,9 @@ def detect(cls, embedder_name: str, device: torch.device):
     def _model_forward_fn(self, input_ids: torch.Tensor, attention_mask: Optional[torch.Tensor] = None) -> torch.Tensor:
         return self._custom_bioengineer.run_model(input_ids, attention_mask)
 
+    def _model_batched_forward_fn(self, input_ids: torch.Tensor, attention_mask: Optional[torch.Tensor] = None) -> torch.Tensor:
+        return self._custom_bioengineer.run_model(input_ids, attention_mask)
```
Comment (SebieF) [inline biotrainer/bioengineer/bioengineer_custom_model.py]: Not sure if that works as expected

File: biotrainer/bioengineer/bioengineer_data_classes.py [hunk 3]
```diff
@@ -264,3 +265,39 @@ def ndcg_score(self):
...
+# ============================================================================
+
+# required for bootstrapping and potentially also for caching and resuming from mid-dataset!
+class ZeroShotContactSingleProtein(BaseModel):  
+    protein_name: str = Field(description="Name of the protein")
+    # TODO: protein sequence, predicted contacts and ground truth not stored to reduce memory usage; add if needed!
+    # protein_sequence: str = Field(description="Protein sequence")
+    # predicted_contact_map: np.ndarray = Field(description="Predicted contact map")
+    # ground_truth_contact_map: np.ndarray = Field(description="Ground truth contact map")
+    precision_scores: Dict[str, float] = Field(description="Precision scores for the protein's predicted contacts")
```
Comment (SebieF) [inline biotrainer/bioengineer/bioengineer_data_classes.py]: Why is it a Dict[str, float] (so what does the str represent)? Maybe add that as a comment.

File: tests/test_bioengineer.py [hunk 4]
```diff
@@ -44,3 +44,36 @@ def test_mutation_parsing(self):
...
+            bio_engineer = BioEngineer.from_baseline(baseline=baseline)
+            self.assertIsNotNone(bio_engineer.model_wrapper, f"Model wrapper for baseline {baseline} is None!")
+            if method not in bio_engineer.model_wrapper.supported_methods():
+                continue
+            results, aggregated_result = bio_engineer.run_contact_dataset(dataset_dir_path=dataset_path, method=method)
+            self.assertNotEqual(results, None)
+            self.assertNotEqual(aggregated_result, None)
+            # TODO: add additional assertions / tests!
+
+    def test_ESM2_8M_UR50D(self):
```
Comment (SebieF) [inline tests/test_bioengineer.py:74]: Please add the annotation
```
    @pytest.mark.skip(reason="Large test that should only be executed on demand")
```

**Review by SebieF — CHANGES_REQUESTED** [review]
Thank you! :)

File: biotrainer/bioengineer/bioengineer.py (hunk 1 above)
Comment (alijosephine) [reply biotrainer/bioengineer/bioengineer.py]: The BioEngineer layer is now agnostic to the directory structure. Instead it expects `dataset_name`, `fasta_file_path`, and `contacts_dir_path` separately.  Is this ok?

File: biotrainer/bioengineer/bioengineer_data_classes.py (hunk 3 above)
Comment (alijosephine) [reply biotrainer/bioengineer/bioengineer_data_classes.py]: str captures the metric name, e.g. "short P@L".

File: biotrainer/bioengineer/bioengineer_custom_model.py (hunk 2 above)
Comment (alijosephine) [reply biotrainer/bioengineer/bioengineer_custom_model.py]: So as we discussed, best way would be to have only `_model_forward_fn`, which would automatically handle batches, and then move the extraction of index `[0]` for single sequence cases to the consumer functions (e.g. `_get_log_probabilities`). 

If this refactoring is to be avoided, and at the same we time do not want to confuse/burden the user with implementing `run_model` correctly for each use-case, then an alternative is to have a separate 
```
    @abstractmethod
    def run_model_batched()
```
(potentially a bit clumsy though!)
What do you suggest?

File: tests/test_bioengineer.py (hunk 4 above)
Comment (alijosephine) [reply tests/test_bioengineer.py:74]: Done.

File: biotrainer/bioengineer/bioengineer.py (hunk 1 above)
Comment (SebieF) [reply biotrainer/bioengineer/bioengineer.py]: Yes :)

File: biotrainer/bioengineer/bioengineer_custom_model.py (hunk 2 above)
Comment (SebieF) [reply biotrainer/bioengineer/bioengineer_custom_model.py]: After taking a closer look, I would suggest to add a default implementation of the `run_model_batched` function to CustomBioEngineerModel:

```python
    def run_model_batched(self, input_ids: torch.Tensor, attention_mask: Optional[torch.Tensor] = None) -> torch.Tensor:
        """
        Run the model on a batch of sequences.

        The default implementation loops over each sequence in the batch and calls
        run_model, then stacks the results. This is for backward compatibility with older versions of bioengineer.
        Override this method for a more efficient native batched forward pass.

        Args:
            input_ids: Token IDs of shape [batch_size, seq_len]
            attention_mask: Optional mask of shape [batch_size, seq_len]

        Returns:
            torch.Tensor of shape [batch_size, seq_len, vocab_size]
        """
        batch_size = input_ids.size(0)
        results = []
        for i in range(batch_size):
            single_ids = input_ids[i].unsqueeze(0)  # [1, seq_len]
            single_mask = attention_mask[i].unsqueeze(0) if attention_mask is not None else None
            logits = self.run_model(single_ids, single_mask)  # [seq_len, vocab_size]
            results.append(logits)
        return torch.stack(results, dim=0)  # [batch_size, seq_len, vocab_size]
```

Then change the wrapper function:
```
    def _model_batched_forward_fn(self, input_ids: torch.Tensor, attention_mask: Optional[torch.Tensor] = None) -> torch.Tensor:
        return self._custom_bioengineer.run_model_batched(input_ids, attention_mask)
```

This will give us the advantage of the batched computation without breaking current code, but allowing overrides of custom implementations for improved performance. What do you think?

Comment (alijosephine) [reply biotrainer/bioengineer/bioengineer_custom_model.py]: This works. But, is not the whole point of having a batched_forward_fn to parallelize and speed up... so if we instead loop over sequences one by one to run the model and then collect results, is it not misleading?

Comment (alijosephine) [reply biotrainer/bioengineer/bioengineer_custom_model.py]: Added this change with notes, for backward compatibility.
